bluesky_kafka/__init__.py

```diff
@@ -305,3 +306,202 @@ def no_work_during_wait():
+            msg.topic(),
+            name,
+            doc,
+        )
+        self.process_document(msg.topic(), name, doc)
+        if name == 'stop':
+            self.finalize_run(doc)
+
+    def _poll(self, work_during_wait):
+        while True:
+            msg = self._consumer.poll(self.polling_duration)
+            if msg is None:
+                # no message was delivered
+                # do some work before polling again
+                work_during_wait()
+            elif msg.error():
+                logger.error("Kafka Consumer error: %s", msg.error())
+            else:
+                try:
+                    self.process(msg)
+                except Exception as exc:
+                    logger.exception(exc)
+
+    def start(self, work_during_wait=None):
+        def no_work_during_wait():
+            # do nothing between message deliveries
+            pass
+
+        if work_during_wait is None:
+            work_during_wait = no_work_during_wait
+
+        if self.closed:
+            raise RuntimeError(
+                "This RemoteDispatcher has already been "
+                "started and interrupted. Create a fresh "
+                f"instance with {repr(self)}"
+            )
+        try:
+            self._poll(work_during_wait=work_during_wait)
+        except Exception:
+            self.stop()
+            raise
+
+    def stop(self):
+        self._consumer.close()
+        self.closed = True
+
+    def finalize_run(self, stop_doc):
+        """
+        Finalize the consumption of the Run.
+        """
+        logger.debug(
+            f"Run consumption complete: {stop_doc['run_start']}, "
+            f"{self._topics}, {self._group_id}"
+        )
+        if self._commit_on_stop_doc:
+            self.commit(asynchronous=False)
+
+
+
+class MongoBlueskyConsumer(BlueskyConsumer):
+    """
+    Subclass of BlueskyConsumer that is specialized for inserting into a mongo
+    database determined by the topic name
+    """
+
+    class SerializerFactory(dict):
+        """
+        Like a defaultdict, but it makes a Serializer based on the
+        key, which in this case is the topic name.
+        """
+        def __init__(self, mongo_uri):
+            self._mongo_uri = mongo_uri
+
+        def get_database(self, topic):
+            return topic.replace('.', ',')
+
+        def __missing__(self, topic):
+            result = self[topic] = Serializer(self._mongo_uri + '/' + self.get_database(topic),
+                                              self._mongo_uri + '/' + self.get_database(topic))
+            return result
+
+    def __init__(self, mongo_uri, *args, **kwargs):
+        self._serializers = self.SerializerFactory(mongo_uri)
+        return super().__init__(*args, **kwargs)
+
+    def process_document(self, topic, name, doc):
+        result_name, result_doc = self._serializers[topic](name, doc)
+        if result_name == 'stop':
+            del self._serializers[topic]
```

Comment:
I think this needs more thought. What is the lifecycle of these serializers? Is it Serializer-per-Run or Serializer-per-topic. As written they are created when we see a new topic (no connection to what Run this is) and destroyed whenever a Run is closed. This would result in a lot of thrashing if interleaved Runs come through. For example, one Run might end up going partially through multiple Serializers.

Given that the number of possible topics is bounded, it might be acceptable to let this Serializers live forever (for the lifecycle of the process). That would be the simplest to implement.

Reply:
Good point.
Interleaved runs will be written to separate partitions.
Most likely this means separate consumers will be handling them.
But if there are less consumers than partitions, then a consumer could be subscribed to two partitions from the same topic. Which could make two runs go through the same serializer.

I think that if we don't delete the Serializers then we would have a memory leak. Over time we could end up with 100 consumers each with 30 Serializers (worst case). Is this okay? We could use another method to delete old serializers.

Reply:
I am not sure we have to call that a leak. Any of the Serializers may (and in practice likely _will_) be used again by the process, and the maximum number we might make has a clear and reasonably low upper bound.

Reply:
Okay, that makes it easy then.

Reply:
Updated